IO/d4est_norms.c

```diff
@@ -246,11 +255,46 @@ d4est_output_norms
     
 }
 
-
-
+void
+d4est_norms_save
+(
+  const char* options_file,
+  const char** u_name,
+  double** u,
+  double** u_compare,
+  d4est_xyz_fcn_t** analytical_solutions,
+  void* analytical_solution_ctx
+)
+{
+  zlog_category_t *c_default = zlog_get_category("d4est_norms");
+  zlog_debug(c_default, "Saving norms...");
+  
+  // First draft using only u_compare
+  int j = -1;
+  while (u_name[++j] != NULL) {
+    int number_of_nodes = sizeof(*u[j]) / sizeof(u[j][0]);
```

Comment:
This doesn't work yet because I don't calculate the length of the array correctly, could you help me out with the C syntax here quickly?

Reply:
the number of nodes on a local processor (aka degrees of freedom, aka size of the field aka number of grid points) can be retrieved by the call 

```C
int num_nodes = d4est_mesh_get_local_nodes(p4est)
```

To get the number of fields (aka variables) sent in, you do:

```C
int num_fields = 0
while (u_name[++j] != NULL){
    num_fields++;
}
```
The above allows us to avoid sending in the actual number of fields we want to compute the norm of.

Then for an example, to loop through the actually fields and compute an Linf norm for each one you might do

```C
double* norms = P4EST_ALLOC(double, num_fields); 
for (int i = 0; i < num_fields; i++){
  double max_error = 0;
  for (int j = 0; j < num_nodes; j++){
    double max_error_temp = fabs(u[i][j] - u_compare[i][j]);
    max_error = (max_error_temp > max_error) ? max_temp_error : max_error;    
   }
  norms[i] = max_error;
}
P4EST_FREE(norms) //if you are done with norms, free it as so
```

You might then call d4est_norms_save if you want to take the norms of two fields, blah1 and blah2, as so
```C
d4est_norms_save(
                               "options.input",
                               (const char* []){"blah1", "blah2", NULL},
                               (double* []){"blah1_ptr", "blah2_ptr"},
                               (double* []){"blah1_compare_ptr" , "blah2_compare_ptr"}  
                               NULL, //no analytical functions here
                               NULL //no analytical function ctx here
                               )
                                
```

Reply:
Great, thanks for your explanations! Indeed this is quite precisely how I implemented the function so far 😃 (Github apparently only shows four lines and I don't see a way to quickly expand the preview, it appears that only when clicking _View changes_ can you see more).

Only what I don't understand yet is how to calculate the number of nodes / degrees of freedom / length of each array in `double **u` without having a pointer to `p4est` available in the function, that I need for `d4est_mesh_get_local_nodes(p4est)`. Usually I would simply calculate the length of the arrays, but I think this is my inexperience with C shining through and from the pointers it's indeed impossible to get the length of the array. Do we need to add an argument `num_nodes` or similar to the function signature to be able to iterate through the arrays?

Reply:
Oops good catch. We need to pass in the p4est structure or the number of nodes. I'm almost certain the p4est structure will be needed somewhere else in the code, but if we also pass in the number of nodes then we don't need to compute it again.

Now that I look at the code more closely though I realize this function requires a lot more context only because of the "energy norm", which is a very specific norm for only a subset of elliptic problems (e.g. it might not be useful for the linear elasticity problem). Maybe we could send in the norms separately to make it completely general and simpler.


```C
d4est_norms_save
(
p4est_t* p4est,
const char ** u_names,
double ** u,
double ** u_compare,  
d4est_xyz_fcn_t* analytical_solutions,
void** analytical_solution_ctx,
const char ** norm_names,
d4est_norm_fcn_t* norm_fcns,
void** norm_fcn_ctx
)
```
where d4est_norm_fcn_t is a function pointer of the form

```C
typedef double
(*d4est_norm_fcn_t)
(
 p4est_t*,
 double *u,
 double *u_compare,
 void* ctx
);
```

and then in d4est_norms.c we will have 

```C
double
d4est_norms_L2(p4est_t* p4est, double* u, double* u_compare, void* ctx){
//..... no ctx needed (just need p4est)
//..... code to compute norm
return norm;
}

double
d4est_norms_Linf(p4est_t* p4est, double* u, double* u_compare, void* ctx){
//..... no ctx needed (just size of arrays, which can be computed using p4est)
//..... code to compute norm
return norm;
}

double
d4est_norms_energy(p4est_t* p4est, double* u, double* u_compare, void* ctx){
d4est_energy_norm_ctx_t* energy_ctx = ctx;
//..... code to compute norm here using energy_ctx
return norm;
}
```

I think this would be more elegant (as elegant as is possible in C anyway) and now we wouldn't really require an options.input entry for the norms (as far as I can see). And yes the sad thing about arrays in C is that the size is not known (unless you wrap the array in your own container, which I haven't done in the code). 

We would then call the function like

```C
d4est_norms_save
(
p4est,
(const char* []){"blah1", "blah2", NULL},
(double* []){blah1_ptr, blah2_ptr},
(double* []){blah1_compare_ptr , blah2_compare_ptr}  
NULL,
NULL,
(const char* []){"L2", "Linf", "energy", NULL},
(d4est_norm_fcn_t []) {d4est_norms_L2, d4est_norms_Linf, d4est_norms_energy},
(void* []){ NULL, NULL, (void*)d4est_norms_energy_ctx}
)
```
and the function would then print out the name of the norm and the value. This is about as general as I can think of at the moment.

Reply:
That sounds great, thanks for working this out 😃 I was actually in the middle of writing comments below but ended up coding instead.

I have a `d4est_norms_save` function including the energy norm working now. As you said, I do indeed pass a whole bunch of context into the function, which is not ideal. I'll restructure it according to your suggestions now. Cheers